File: jas_ocaml/lib/interpreter/yaml_dialog_view.ml
(** YAML-interpreted dialog component for GTK3.

    Renders a modal dialog from workspace YAML definitions, reusing the
    existing element renderer for the content tree. *)

(** Dialog state. *)
type dialog_state = {
  id : string;
  state : (string * Yojson.Safe.t) list;
  params : (string * Yojson.Safe.t) list;
}

(** Read the live dialog state list (forwards to [Dialog_global]). *)
let current_dialog_state () : (string * Yojson.Safe.t) list =
  Dialog_global.read_state ()

(** Set [key] in the live dialog state. *)
let set_dialog_field (key : string) (value : Yojson.Safe.t) : unit =
  Dialog_global.set_field key value

(** Close the active dialog widget. *)
let close_current_dialog () : unit = Dialog_global.close ()

(** Build a fresh evaluation context against the live dialog state. *)
let current_dialog_ctx () : Yojson.Safe.t =
  !Dialog_global.current_build_ctx ()

(** Open a dialog by ID, initializing its state from the workspace
    definition. ``outer_scope`` exposes top-level keys (like
    ``active_document``) to init expressions. *)
let open_dialog ?(outer_scope : (string * Yojson.Safe.t) list = [])
    (dialog_id : string)
    (raw_params : (string * Yojson.Safe.t) list)
    (live_state : (string * Yojson.Safe.t) list) : dialog_state option =
  match Workspace_loader.load () with
  | None -> None
  | Some ws ->
    match Workspace_loader.dialog ws dialog_id with
    | None -> None
    | Some (`Assoc dlg_def) ->
      let defaults = match List.assoc_opt "state" dlg_def with
        | Some defs -> Effects.state_defaults defs
        | None -> []
      in
      let state_ctx = `Assoc (
        ("state", `Assoc live_state) :: outer_scope
      ) in
      (* Resolve params only when the value is an expression
         (contains a dot — identifier path access). Bare-word values
         like [target: fill] are already literal at this point — the
         caller's click-behavior resolver attempted to evaluate them
         and fell back to the raw string when the identifier wasn't
         bound. Re-evaluating "fill" here would return Null (Expr_eval
         treats undefined identifiers as Null) and the dialog's
         [if param.target == "fill"] branch would silently fall
         through to the else branch. *)
      let resolved_params = List.map (fun (k, v) ->
        match v with
        | `String s when String.contains s '.' ->
          let result = Expr_eval.evaluate s state_ctx in
          (k, Effects.value_to_json result)
        | _ -> (k, v)
      ) raw_params in
      let dialog_state = ref defaults in
      let init_ctx_for () : Yojson.Safe.t =
        `Assoc (
          ("state", `Assoc live_state) ::
          ("dialog", `Assoc !dialog_state) ::
          ("param", `Assoc resolved_params) ::
          outer_scope)
      in
      (match List.assoc_opt "init" dlg_def with
       | Some (`Assoc init_map) ->
         let deferred = ref [] in
         List.iter (fun (key, expr) ->
           let expr_str = match expr with `String s -> s | _ -> "" in
           if (try ignore (Str.search_forward
                             (Str.regexp_string "dialog.") expr_str 0); true
               with Not_found -> false) then
             deferred := (key, expr) :: !deferred
           else begin
             let result = Expr_eval.evaluate expr_str (init_ctx_for ()) in
             dialog_state := (key, Effects.value_to_json result) ::
               List.filter (fun (k, _) -> k <> key) !dialog_state
           end
         ) init_map;
         List.iter (fun (key, expr) ->
           let expr_str = match expr with `String s -> s | _ -> "" in
           let result = Expr_eval.evaluate expr_str (init_ctx_for ()) in
           dialog_state := (key, Effects.value_to_json result) ::
             List.filter (fun (k, _) -> k <> key) !dialog_state
         ) (List.rev !deferred)
       | _ -> ());
      Some {
        id = dialog_id;
        state = !dialog_state;
        params = resolved_params;
      }
    | _ -> None

(** Dispatch a YAML action by name (forwards to [Dialog_global]). *)
let dispatch_action = Dialog_global.dispatch_action

(** Show a YAML dialog as a GTK modal dialog. *)
let show_dialog ?(parent : GWindow.window option)
    ?(outer_scope : (string * Yojson.Safe.t) list = [])
    (ds : dialog_state) : unit =
  match Workspace_loader.load () with
  | None -> ()
  | Some ws ->
    match Workspace_loader.dialog ws ds.id with
    | Some (`Assoc dlg_def) ->
      let summary = match List.assoc_opt "summary" dlg_def with
        | Some (`String s) -> s
        | _ -> ds.id
      in
      let width = match List.assoc_opt "width" dlg_def with
        | Some (`Int w) -> w
        | _ -> 400
      in
      let dialog = GWindow.dialog
        ?parent
        ~title:summary
        ~width
        ~modal:true
        ~destroy_with_parent:true
        () in
      let vbox = dialog#vbox in
      (* Theme the modal dialog window with the active appearance pane
         background so it matches the dock / panels (GTK otherwise draws a
         default LIGHT dialog). GtkBox / GtkLabel / GtkFrame are no-window
         widgets that show the window background through; inputs / buttons /
         labels theme their own colors via the render_element theme hooks.
         Mirrors the flyout background in show_nonmodal_dialog. *)
      (let dialog_bg = !(Yaml_panel_view.dialog_pane_bg_hook) () in
       let dialog_css = new GObj.css_provider (GtkData.CssProvider.create ()) in
       dialog_css#load_from_data (Printf.sprintf
         "window, dialog, box, grid { background-color: %s; }" dialog_bg);
       dialog#misc#style_context#add_provider dialog_css 600;
       vbox#misc#style_context#add_provider dialog_css 600);
      let state_defaults = Workspace_loader.state_defaults ws in
      let icons = Workspace_loader.icons ws in
      let live_state = ref ds.state in
      (* Parse state get/set props from the YAML so dialog widgets'
         write-backs run the declared setter (e.g. color_picker's
         h-setter rebuilds [color] from the new h + existing s/b).
         Without props, channel edits would be silent no-ops on the
         canonical color and the preview swatch / hue bar wouldn't
         follow the typed value. *)
      let props = match List.assoc_opt "state" dlg_def with
        | Some (`Assoc state_defs) ->
          List.filter_map (fun (key, def) ->
            match def with
            | `Assoc fields ->
              let get = match List.assoc_opt "get" fields with
                | Some (`String s) -> Some s | _ -> None in
              let set = match List.assoc_opt "set" fields with
                | Some (`String s) -> Some s | _ -> None in
              if get = None && set = None then None
              else Some (key, { Dialog_global.prop_get = get; prop_set = set })
            | _ -> None
          ) state_defs
        | _ -> []
      in
      Dialog_global.current_state := Some live_state;
      Dialog_global.current_id := Some ds.id;
      Dialog_global.current_outer_scope := outer_scope;
      Dialog_global.current_props := props;
      Dialog_global.clear_state_change_listeners ();
      Dialog_global.current_close := (fun () -> dialog#destroy ());
      Dialog_global.current_build_ctx := (fun () ->
        `Assoc (
          ("state", `Assoc state_defaults) ::
          ("dialog", `Assoc (Dialog_global.read_state ())) ::
          ("param", `Assoc ds.params) ::
          ("icons", icons) ::
          outer_scope
        ));
      let ctx = !Dialog_global.current_build_ctx () in
      (match List.assoc_opt "content" dlg_def with
       | Some content ->
         Yaml_panel_view.render_element
           ~packing:(vbox#pack ~expand:false)
           ~ctx content
       | None -> ());
      dialog#show ();
      ignore (dialog#run ());
      (try dialog#destroy () with _ -> ());
      Dialog_global.current_state := None;
      Dialog_global.current_id := None;
      Dialog_global.current_outer_scope := [];
      Dialog_global.current_props := [];
      Dialog_global.clear_state_change_listeners ();
      Dialog_global.current_close := (fun () -> ());
      Dialog_global.current_build_ctx := (fun () -> `Assoc [])
    | _ -> ()

(** Show a [modal: false] YAML dialog as a non-blocking GTK flyout.

    Unlike [show_dialog] (which is modal and runs a blocking
    [dialog#run] loop), this opens an undecorated top-level window
    that does NOT grab the main loop — so the canvas / toolbar stay
    live and the popup dismisses itself on a click outside or when an
    item's [close_dialog] effect fires. Used by the toolbar's
    long-press tool-alternates flyout.

    The window is positioned at the current pointer location so it
    appears next to the slot button the user is holding. The same
    [Dialog_global] wiring [show_dialog] installs is reused, so an
    item button's [set: { active_tool: ... }] + [close_dialog]
    behavior dispatches through the identical render-button path. *)
let show_nonmodal_dialog (ds : dialog_state) : unit =
  match Workspace_loader.load () with
  | None -> ()
  | Some ws ->
    match Workspace_loader.dialog ws ds.id with
    | Some (`Assoc dlg_def) ->
      (* If a prior flyout is still open, dismiss it first so two
         long-presses don't stack windows. *)
      Dialog_global.close ();
      let win = GWindow.window
        ~type_hint:`POPUP_MENU
        ~decorated:false
        ~resizable:false
        ~show:false
        () in
      win#set_skip_taskbar_hint true;
      win#set_skip_pager_hint true;
      (* Accept + grab focus on map so the focus-out dismissal below is
         reliable: the popup holds focus, and the user's next click
         elsewhere moves focus away and closes it (menu semantics). *)
      win#set_accept_focus true;
      win#set_focus_on_map true;
      (* Anchor the flyout at the pointer (the slot button the user is
         long-pressing). [window_at_pointer] gives the window under the
         cursor plus pointer-relative coords; that window's root origin
         converts them to absolute screen coords. Offset a touch so the
         popup sits below-right of the cursor rather than under it.
         Best-effort: a centered popup if the pointer can't be read. *)
      (try
        match Gdk.Display.window_at_pointer () with
        | Some (gwin, rx, ry) ->
          let (ox, oy) = Gdk.Window.get_origin gwin in
          win#move ~x:(ox + rx + 4) ~y:(oy + ry + 4)
        | None -> ()
      with _ -> ());
      (* Dark popup chrome so the flyout matches the dark toolbar/menu:
         GTK's default popup background is light, which makes the
         #cccccc-tinted tool icons read as gray. A dark bg makes the
         same icons pop bright/white like they do on the toolbar. Use
         [`NONE] shadow on the frame so GTK's light 3D bevel doesn't
         fight the dark fill, and paint a subtle border + radius via a
         CSS provider (mirrors how menubar.ml builds + adds a provider).
         The bg color comes from [dialog_bg_hook] (wired in main.ml to
         the active appearance's dark pane background). *)
      let frame = GBin.frame ~shadow_type:`NONE ~packing:win#add () in
      let vbox = GPack.vbox ~packing:frame#add () in
      let flyout_bg = !(Yaml_panel_view.dialog_bg_hook) () in
      let flyout_css = new GObj.css_provider (GtkData.CssProvider.create ()) in
      flyout_css#load_from_data (Printf.sprintf
        "window, frame, frame > border, box { background-color: %s; } \
         frame, frame > border { border: 1px solid #5a5a5a; border-radius: 4px; padding: 2px; }"
        flyout_bg);
      win#misc#style_context#add_provider flyout_css 600;
      frame#misc#style_context#add_provider flyout_css 600;
      vbox#misc#style_context#add_provider flyout_css 600;
      let state_defaults = Workspace_loader.state_defaults ws in
      let icons = Workspace_loader.icons ws in
      let live_state = ref ds.state in
      let props = match List.assoc_opt "state" dlg_def with
        | Some (`Assoc state_defs) ->
          List.filter_map (fun (key, def) ->
            match def with
            | `Assoc fields ->
              let get = match List.assoc_opt "get" fields with
                | Some (`String s) -> Some s | _ -> None in
              let set = match List.assoc_opt "set" fields with
                | Some (`String s) -> Some s | _ -> None in
              if get = None && set = None then None
              else Some (key, { Dialog_global.prop_get = get; prop_set = set })
            | _ -> None
          ) state_defs
        | _ -> []
      in
      Dialog_global.current_state := Some live_state;
      Dialog_global.current_id := Some ds.id;
      Dialog_global.current_outer_scope := [];
      Dialog_global.current_props := props;
      Dialog_global.clear_state_change_listeners ();
      (* close_dialog / dismiss_dialog and the outside-click handler
         all route through this. Tear down the Dialog_global wiring on
         destroy so a closed flyout leaves no stale state behind. *)
      let teardown () =
        Dialog_global.current_state := None;
        Dialog_global.current_id := None;
        Dialog_global.current_outer_scope := [];
        Dialog_global.current_props := [];
        Dialog_global.clear_state_change_listeners ();
        Dialog_global.current_close := (fun () -> ());
        Dialog_global.current_build_ctx := (fun () -> `Assoc [])
      in
      Dialog_global.current_close := (fun () ->
        teardown ();
        (try win#destroy () with _ -> ()));
      Dialog_global.current_build_ctx := (fun () ->
        `Assoc (
          ("state", `Assoc state_defaults) ::
          ("dialog", `Assoc (Dialog_global.read_state ())) ::
          ("param", `Assoc ds.params) ::
          ("icons", icons) :: []
        ));
      let ctx = !Dialog_global.current_build_ctx () in
      (* Render the flyout's tool icons larger than the 20px panel
         default — close to the 32px toolbar so the alternates read as
         prominent tool buttons, not shrunken glyphs. The items in
         tool_alternates.yaml declare no [style.size], so they pick up
         this scoped override. Restored to [None] in a finally so panels
         and modal dialogs keep their 20px default even if render raises. *)
      Yaml_panel_view.nonmodal_icon_size := Some 28;
      (match List.assoc_opt "content" dlg_def with
       | Some content ->
         (match Yaml_panel_view.render_element
            ~packing:(vbox#pack ~expand:false)
            ~ctx content
          with
          | () -> Yaml_panel_view.nonmodal_icon_size := None
          | exception e ->
            Yaml_panel_view.nonmodal_icon_size := None;
            raise e)
       | None -> Yaml_panel_view.nonmodal_icon_size := None);
      (* Dismiss on focus-out (a click anywhere outside the popup moves
         focus away, so the flyout closes — matching menu semantics:
         the next click either lands on an item (running its
         close_dialog) or dismisses the popup). *)
      ignore (win#event#connect#focus_out ~callback:(fun _ ->
        Dialog_global.close (); false));
      win#show ();
      win#present ()
    | _ -> ()

File: JasSwift/Sources/Interpreter/YamlDialogView.swift
/// YAML-interpreted dialog component for SwiftUI.
///
/// Renders a modal dialog from workspace YAML definitions, reusing the
/// existing YamlElementView for the content tree.

import SwiftUI

/// Dialog state held in a binding.
struct YamlDialogState {
    /// Dialog ID (matches a key in workspace dialogs).
    var id: String
    /// Dialog-local state values.
    var state: [String: Any]
    /// Parameters passed when the dialog was opened.
    var params: [String: Any]
    /// Popover anchor in the window's `.global` coordinate space —
    /// the page coordinates of the press that opened the dialog.
    /// Mirrors Rust's ``DialogState.anchor`` (page coords captured at
    /// the slot button's mouse_down, threaded through the long-press
    /// timer). Only consulted for non-modal dialogs; modal dialogs
    /// stay centered regardless. Default nil → centered fallback,
    /// matching Rust's ``anchor: None`` branch.
    var anchor: CGPoint? = nil
}

/// Open a dialog by ID, initializing its state from the workspace
/// definition. Back-compatible shim over openYamlDialogWithOuter
/// without a cross-scope namespace — dialogs whose init expressions
/// reference `panel.*` or `active_document.*` (Artboard Options)
/// must call the `_withOuter` variant directly.
func openYamlDialog(
    dialogId: String,
    rawParams: [String: Any],
    liveState: [String: Any]
) -> YamlDialogState? {
    openYamlDialogWithOuter(
        dialogId: dialogId,
        rawParams: rawParams,
        liveState: liveState,
        outerScope: [:]
    )
}

/// As ``openYamlDialog`` but threads an ``outerScope`` dictionary whose
/// top-level keys (e.g. ``panel``, ``active_document``) are visible to
/// init expressions alongside ``state``, ``dialog``, ``param``. Required
/// by the Artboard Options Dialogue whose init expressions call
/// ``filter(active_document.artboards, ...)`` and
/// ``panel.reference_point``.
func openYamlDialogWithOuter(
    dialogId: String,
    rawParams: [String: Any],
    liveState: [String: Any],
    outerScope: [String: Any]
) -> YamlDialogState? {
    guard let ws = WorkspaceData.load() else { return nil }
    guard let dlgDef = ws.dialog(dialogId) else { return nil }

    // Extract state defaults
    var defaults: [String: Any] = [:]
    if let stateDefs = dlgDef["state"] as? [String: Any] {
        for (key, defn) in stateDefs {
            if let d = defn as? [String: Any] {
                defaults[key] = d["default"]
            } else {
                defaults[key] = defn
            }
        }
    }

    // Resolve param expressions against current state
    var resolvedParams: [String: Any] = [:]
    let stateCtx: [String: Any] = ["state": liveState]
    for (k, v) in rawParams {
        if let exprStr = v as? String {
            let result = evaluate(exprStr, context: stateCtx)
            resolvedParams[k] = valueToAnyDlg(result)
        } else {
            resolvedParams[k] = v
        }
    }

    var dialogState = defaults

    // Evaluate init expressions (two-pass for dict order independence).
    // Both passes get outerScope so non-dialog-referencing init exprs
    // can still read `active_document.*` and `panel.*` (matches Rust's
    // build_dialog_outer_scope behavior; without this fix any init
    // that reads e.g. `active_document.print_preferences.copies`
    // would silently resolve to null and the dialog would open with
    // the YAML state defaults instead of the persisted document
    // values).
    if let initMap = dlgDef["init"] as? [String: Any] {
        var deferred: [(String, Any)] = []
        for (key, expr) in initMap {
            let exprStr = expr as? String ?? ""
            if exprStr.contains("dialog.") {
                deferred.append((key, expr))
            } else {
                var initCtx: [String: Any] = outerScope
                initCtx["state"] = liveState
                initCtx["dialog"] = dialogState
                initCtx["param"] = resolvedParams
                let result = evaluate(exprStr, context: initCtx)
                dialogState[key] = valueToAnyDlg(result)
            }
        }
        for (key, expr) in deferred {
            let exprStr = expr as? String ?? ""
            var initCtx: [String: Any] = outerScope
            initCtx["state"] = liveState
            initCtx["dialog"] = dialogState
            initCtx["param"] = resolvedParams
            let result = evaluate(exprStr, context: initCtx)
            dialogState[key] = valueToAnyDlg(result)
        }
    }

    return YamlDialogState(
        id: dialogId,
        state: dialogState,
        params: resolvedParams
    )
}

/// Build a ``YamlDialogState`` snapshot from a StateStore that has a
/// dialog open. Returns nil when the store has no active dialog.
///
/// Used by the action → overlay bridge: after running YAML effects,
/// if ``store.getDialogId()`` transitioned from nil to an id, the
/// dispatcher calls this helper and assigns the result to the
/// SwiftUI `yamlDialogState` binding. The Artboard Options Dialogue
/// is opened this way (via the `open_dialog` effect); the color
/// picker uses the direct `openYamlDialog` path instead.
func yamlDialogStateFromStore(_ store: StateStore) -> YamlDialogState? {
    guard let id = store.getDialogId() else { return nil }
    return YamlDialogState(
        id: id,
        state: store.getDialogState(),
        params: store.getDialogParams() ?? [:]
    )
}

/// Build the evaluation context surfaced to a dialog's content-tree
/// renderer. Outer-scope keys (``panel``, ``active_document``) are
/// merged first; the dialog's own ``dialog`` / ``param`` / ``state``
/// / ``icons`` keys win on collision so dialog-local state is never
/// stomped by a like-named outer key.
///
/// Exposed at file scope so Phase F tests can verify merge order
/// without standing up a SwiftUI view.
func buildDialogEvalContext(
    state: [String: Any],
    params: [String: Any],
    outer: [String: Any]
) -> [String: Any] {
    var ctx: [String: Any] = outer
    ctx["dialog"] = state
    ctx["param"] = params
    if let ws = WorkspaceData.load() {
        ctx["state"] = ws.stateDefaults()
        ctx["icons"] = ws.icons()
    }
    return ctx
}

/// Dialog ctx that also resolves YAML-defined getters via the store.
///
/// Color picker H/S/B/R/G/B/C/M/Y/K/hex are declared with `get:`
/// expressions in the YAML state block; they aren't stored in the
/// dialog map. Plain `state["h"]` lookup returns nil, so a widget
/// bound to `dialog.h` would render 0. This variant computes each
/// getter via `StateStore.getDialogWithOuter` and merges the values
/// into the `dialog` map so renderer expressions see them.
func buildDialogEvalContextWithGetters(
    store: StateStore?,
    state: [String: Any],
    params: [String: Any],
    outer: [String: Any]
) -> [String: Any] {
    var dialog = state
    if let store = store, let dlgId = store.getDialogId(),
       let ws = WorkspaceData.load(),
       let dlgDef = ws.dialog(dlgId),
       let stateDefs = dlgDef["state"] as? [String: Any]
    {
        for (key, defn) in stateDefs {
            guard let d = defn as? [String: Any], d["get"] != nil else { continue }
            if let v = store.getDialogWithOuter(key, outer: outer) {
                dialog[key] = v
            }
        }
    }
    return buildDialogEvalContext(state: dialog, params: params, outer: outer)
}

/// Convert a Value to Any? for storage.
private func valueToAnyDlg(_ v: Value) -> Any? {
    switch v {
    case .null: return nil
    case .bool(let b): return b
    case .number(let n):
        if n == Double(Int(n)) { return Int(n) }
        return n
    case .string(let s): return s
    case .color(let c): return c
    case .list(let l): return l.map { $0.value }
    case .path(let p): return ["__path__": p]
    case .closure: return v  // keep as Value for closure dispatch
    }
}

/// SwiftUI view that renders a YAML dialog as a modal overlay.
///
/// ``outerScope`` is a closure evaluated on each render that returns a
/// dictionary merged under the dialog's own ``dialog`` / ``param`` keys.
/// It exposes ``panel`` + ``active_document`` namespaces to render-time
/// bind expressions — e.g. the Artboard Options Dialogue uses
/// ``bind.disabled: active_document.artboards_count <= 1`` on its
/// Delete button and interpolates ``{{active_document.artboards_count}}``
/// into a "Artboards: N" label. Default: an empty closure, matching
/// the back-compat path for dialogs (like color_picker) that read only
/// ``dialog.*`` / ``state.*``.
struct YamlDialogOverlay: View {
    @Binding var dialogState: YamlDialogState?
    let theme: Theme
    var outerScope: () -> [String: Any] = { [:] }
    /// Active Model used for dialog-body widget dispatch. Needs to
    /// be the same Model whose StateStore holds the dialog state
    /// that this overlay is mirroring so that ``close_dialog``
    /// effects (e.g. via ``artboard_options_confirm``'s tail) zero
    /// the store we're watching.
    var model: Model? = nil
    /// Fired when the overlay is dismissed from the UI (X button or
    /// backdrop tap). Callers pair it with `model.stateStore.closeDialog()`
    /// so the SwiftUI binding and the store's dialog tracker stay in
    /// sync — otherwise a subsequent `open_dialog: <same-id>` effect
    /// would be a no-op transition and the bridge would miss it.
    var onDismiss: (() -> Void)? = nil

    var body: some View {
        if let ds = dialogState {
            if !isModal(ds), let anchor = ds.anchor {
                // Popover branch (non-modal + anchored): transparent
                // full-screen backdrop (click-outside dismiss) + a
                // container pinned next to the press location. Mirrors
                // Rust's `!is_modal && anchor==Some` branch
                // (dialog_view.rs:452-457): position:absolute; left/top
                // at the cursor's page coords, no flex centering, no
                // title bar. The slot button is ~28pt wide; offset the
                // popover a touch to the right of the press so it sits
                // beside the button rather than under the cursor.
                anchoredPopover(ds, anchor: anchor)
            } else {
                // Modal (and non-modal-without-anchor) branch: dimmed,
                // flex-centered overlay exactly as before. Color
                // picker / tool-options / print / artboard options all
                // land here and are visually unchanged.
                centeredModal(ds)
            }
        }
    }

    /// Read the dialog spec's `modal` flag, defaulting to true (matches
    /// Rust's ``dlg_def.get("modal")...unwrap_or(true)``). Only
    /// ``modal: false`` dialogs (the toolbar tool-alternates flyouts)
    /// are eligible for at-cursor placement.
    private func isModal(_ ds: YamlDialogState) -> Bool {
        guard let ws = WorkspaceData.load(),
              let dlgDef = ws.dialog(ds.id),
              let m = dlgDef["modal"] as? Bool else { return true }
        return m
    }

    /// Centered modal/backdrop overlay (the pre-existing presentation).
    @ViewBuilder
    private func centeredModal(_ ds: YamlDialogState) -> some View {
        ZStack {
            // Backdrop
            SwiftUI.Color.black.opacity(0.15)
                .ignoresSafeArea()
                .onTapGesture {
                    dialogState = nil
                    onDismiss?()
                }

            // Dialog container
            VStack(spacing: 0) {
                // Title bar
                titleBar(ds)

                // Body
                dialogBody(ds)
            }
            .background(SwiftUI.Color(nsColor: theme.paneBg))
            .cornerRadius(8)
            .overlay(
                RoundedRectangle(cornerRadius: 8)
                    .stroke(SwiftUI.Color(nsColor: theme.border), lineWidth: 1)
            )
            .shadow(color: .black.opacity(0.25), radius: 16, x: 0, y: 8)
            .frame(maxWidth: dialogWidth(ds))
            // Without this the outer VStack expands vertically to
            // fill the ZStack — Spacer()s in the tabs renderer
            // (used for the left rail and the content column)
            // push the dialog to full window height. Forcing
            // intrinsic vertical sizing fits the dialog to its
            // content; the inner Spacers still let tab content
            // top-align within the available rail height.
            .fixedSize(horizontal: false, vertical: true)
        }
    }

    /// Non-modal popover pinned at `anchor` (window `.global` coords).
    /// The transparent backdrop fills the window and dismisses on tap
    /// (mirrors Rust's transparent `inset:0` backdrop whose onmousedown
    /// clears the dialog). The container is placed with `.position` in
    /// the SAME `.global` coordinate space the anchor was captured in,
    /// so left/top math matches Rust's `position:absolute; left/top`.
    @ViewBuilder
    private func anchoredPopover(_ ds: YamlDialogState, anchor: CGPoint) -> some View {
        // `anchor` is the press point in the shared "jasRoot" coordinate
        // space. Convert it into THIS overlay container's local space
        // (via the container's own frame in "jasRoot") and pin the
        // popover's TOP-LEFT there with `.offset` from a topLeading
        // ZStack — so the flyout lands exactly at the press point no
        // matter where the overlay sits in the window. A small nudge
        // keeps the first item clear of the release point. Matches
        // Rust's at-cursor `position:absolute; left/top` flyout.
        let dx: CGFloat = 8
        let dy: CGFloat = 8
        GeometryReader { geo in
            let o = geo.frame(in: .named("jasRoot")).origin
            ZStack(alignment: .topLeading) {
                // Transparent full-area click-outside dismiss target.
                SwiftUI.Color.clear
                    .contentShape(Rectangle())
                    .onTapGesture {
                        dialogState = nil
                        onDismiss?()
                    }
                popoverContainer(ds)
                    .offset(x: anchor.x - o.x + dx, y: anchor.y - o.y + dy)
            }
        }
    }

    /// The popover's content container: bare (no title bar), shadowed,
    /// width-constrained. Reuses ``dialogBody`` so flyout items render
    /// through the same YamlElementView path as every other dialog.
    @ViewBuilder
    private func popoverContainer(_ ds: YamlDialogState) -> some View {
        VStack(spacing: 0) {
            dialogBody(ds)
        }
        .background(SwiftUI.Color(nsColor: theme.paneBg))
        .cornerRadius(8)
        .overlay(
            RoundedRectangle(cornerRadius: 8)
                .stroke(SwiftUI.Color(nsColor: theme.border), lineWidth: 1)
        )
        .shadow(color: .black.opacity(0.25), radius: 16, x: 0, y: 8)
        // Size the flyout to its intrinsic content (a compact tool list)
        // rather than the 500pt default dialog width. The alternates
        // items are width:"100%", so a wide frame made the popover a
        // 500px box whose bulk read as ~300px to the right of the cursor
        // even though its top-left was pinned at the press point.
        .fixedSize()
    }

    private func titleBar(_ ds: YamlDialogState) -> some View {
        let ws = WorkspaceData.load()
        let dlgDef = ws?.dialog(ds.id)
        let summary = (dlgDef?["summary"] as? String) ?? ds.id

        return HStack {
            SwiftUI.Text(summary)
                .font(.system(size: 13, weight: .medium))
                .foregroundColor(SwiftUI.Color(nsColor: theme.titleBarText))

            Spacer()

            Button(action: {
                dialogState = nil
                onDismiss?()
            }) {
                SwiftUI.Text("\u{00d7}")
                    .font(.system(size: 16))
                    .foregroundColor(SwiftUI.Color(nsColor: theme.textDim))
            }
            .buttonStyle(.plain)
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 8)
        .background(SwiftUI.Color(nsColor: theme.titleBarBg))
    }

    @ViewBuilder
    private func dialogBody(_ ds: YamlDialogState) -> some View {
        let ws = WorkspaceData.load()
        let dlgDef = ws?.dialog(ds.id)
        let content = dlgDef?["content"] as? [String: Any]

        if let content = content {
            let ctx = buildDialogEvalContextWithGetters(
                store: model?.stateStore,
                state: ds.state,
                params: ds.params,
                outer: outerScope()
            )
            YamlElementView(
                element: content,
                context: ctx,
                model: model,
                onWidgetAction: handleDialogWidgetAction,
                theme: theme,
                onDialogWrite: handleDialogStateWrite,
                onStoreDialogClosed: { dialogState = nil }
            )
                .foregroundColor(SwiftUI.Color(nsColor: theme.text))
                .padding(4)
        }
    }

    /// Receive a write from a dialog-body widget bound to ``dialog.X``.
    /// Updates the underlying StateStore (which runs the prop's set
    /// lambda when present) and re-syncs the SwiftUI ``dialogState``
    /// binding from the post-write store snapshot. The earlier
    /// implementation also wrote `value` into `ds.state[key]`
    /// directly, which bypassed the setter — fields with get/set
    /// (color picker H/S/B/R/G/B/C/M/Y/K) ended up with the typed
    /// value in `state[key]` but stale derived values everywhere
    /// else, and the field snapped back to its computed-from-color
    /// value on the next render.
    private func handleDialogStateWrite(_ key: String, _ value: Any?) {
        guard let store = model?.stateStore else {
            guard var ds = dialogState else { return }
            ds.state[key] = value
            dialogState = ds
            return
        }
        store.setDialog(key, value)
        // Color picker "Only Web Colors": when toggled on, snap each
        // RGB channel to multiples of 51 (0/51/102/153/204/255). The
        // r/g/bl setters route through the YAML lambdas which rebuild
        // dialog.color, so subsequent reads of h/s/b/r/g/bl/c/m/y/k/hex
        // see the snapped value.
        if key == "web_only", let on = value as? Bool, on {
            func snap(_ v: Int) -> Int {
                let n = (Double(v) / 51.0).rounded() * 51.0
                return min(max(Int(n), 0), 255)
            }
            func ival(_ k: String) -> Int {
                if let any = store.getDialog(k) {
                    if let i = any as? Int { return i }
                    if let d = any as? Double { return Int(d.rounded()) }
                }
                return 0
            }
            store.setDialog("r", snap(ival("r")))
            store.setDialog("g", snap(ival("g")))
            store.setDialog("bl", snap(ival("bl")))
        }
        if var ds = dialogState {
            ds.state = store.getDialogState()
            dialogState = ds
        }
    }

    /// Dispatch a dialog-body widget-level ``action:`` click. Params
    /// are already resolved against the dialog ctx by YamlElementView.
    ///
    /// Two action shapes:
    /// - ``dismiss_dialog`` — close the store and zero the binding.
    ///   Matches the Python YamlDialogView._dispatch_dialog_action
    ///   special case (Cancel button idiom).
    /// - any other action name — route through
    ///   ``LayersPanel.dispatchYamlAction`` with the resolved params.
    ///   If the action's effects include ``close_dialog``, the store's
    ///   dialog id becomes nil; we mirror that by zeroing the binding.
    private func handleDialogWidgetAction(_ actionName: String,
                                           _ params: [String: Any]) {
        if actionName == "dismiss_dialog" {
            model?.stateStore.closeDialog()
            dialogState = nil
            return
        }
        guard let m = model else { return }
        let abSel = (m.stateStore.getPanelState("artboards")["artboards_panel_selection"] as? [Any])?
            .compactMap { $0 as? String } ?? []
        LayersPanel.dispatchYamlAction(
            actionName, model: m,
            artboardsPanelSelection: abSel,
            params: params
        )
        // The action's effects may have run close_dialog — sync the
        // SwiftUI binding so the overlay dismisses.
        if m.stateStore.getDialogId() == nil {
            dialogState = nil
        }
    }

    private func dialogWidth(_ ds: YamlDialogState) -> CGFloat {
        guard let ws = WorkspaceData.load(),
              let dlgDef = ws.dialog(ds.id),
              let w = dlgDef["width"] as? NSNumber else { return 500 }
        return CGFloat(w.doubleValue)
    }
}

